server/src/main/java/io/crate/replication/logical/action/GetStoreMetadataAction.java

```diff
@@ -0,0 +1,159 @@
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+
+public class GetStoreMetadataAction extends ActionType<GetStoreMetadataAction.Response> {
+
+    public static final String NAME = "internal:crate:replication/logical/store/file_metadata";
+    public static final GetStoreMetadataAction INSTANCE = new GetStoreMetadataAction();
+
+    public GetStoreMetadataAction() {
+        super(NAME);
+    }
+
+    @Override
+    public Writeable.Reader<Response> getResponseReader() {
+        return Response::new;
+    }
+
+    @Singleton
+    public static class TransportAction extends TransportSingleShardAction<Request, Response> {
+
+        private static final Logger LOGGER = Loggers.getLogger(TransportAction.class);
+
+        private final PublisherRestoreService publisherRestoreService;
+
+        @Inject
+        public TransportAction(ThreadPool threadPool,
+                               ClusterService clusterService,
+                               TransportService transportService,
+                               PublisherRestoreService publisherRestoreService,
+                               IndexNameExpressionResolver indexNameExpressionResolver) {
+            super(
+                NAME,
+                threadPool,
+                clusterService,
+                transportService,
+                indexNameExpressionResolver,
+                Request::new,
+                ThreadPool.Names.LOGICAL_REPLICATION
+            );
+            this.publisherRestoreService = publisherRestoreService;
+            TransportActionProxy.registerProxyAction(transportService, NAME, Response::new);
+        }
+
+        @Override
+        protected Response shardOperation(Request request,
+                                          ShardId shardId) throws IOException {
+            if (LOGGER.isTraceEnabled()) {
+                LOGGER.trace("Processing request: " + request.toString());
+            }
+            var metadataSnapshot = publisherRestoreService.createRestoreContext(
+                request.restoreUUID(),
+                request
+            ).metadataSnapshot();
+            if (LOGGER.isTraceEnabled()) {
+                LOGGER.trace("MetadataSnapshot: {}", metadataSnapshot.asMap());
+            }
+            return new Response(metadataSnapshot);
+        }
+
+        @Override
+        protected Writeable.Reader<Response> getResponseReader() {
+            return Response::new;
+        }
+
+        @Override
+        protected boolean resolveIndex(Request request) {
+            return true;
+        }
+
+        @Nullable
+        @Override
+        protected ShardsIterator shards(ClusterState state,
+                                        InternalRequest request) {
+            return state.routingTable().shardRoutingTable(request.request().publisherShardId()).primaryShardIt();
+        }
+    }
+
+    public static class Request extends RestoreShardRequest<Request> {
```

Comment:
There are 3 classes extending `RestoreShardRequest` and each is just delegating to the base class. Could we make `RestoreShardRequest` non-abstract and use it directly?

Reply:
`GetFileChunkAction.Request` extends from it while adding variables, so unfortunately we can't use it directly.